### PR TITLE
Make expectation op accept SimulatesExpectationValues

### LINKED ISSUE #490: Expectation layer won't work with remote simulators
Currently, the `Expectation` layer requires a `cirq.SimulatesFinalState`, see [this line](https://github.com/tensorflow/quantum/blob/52ccd823a7062f318304e5964b55ae521038b3f7/tensorflow_quantum/core/ops/circuit_execution_ops.py#L125).  Instead, it should be updated to directly call the methods in the new Cirq class [`cirq.SimulatesExpectationValues`](https://github.com/quantumlib/Cirq/blob/8cf782554adbafed724987de3067de7ca565fa0c/cirq/sim/simulator.py#L202).

### PR DESCRIPTION
**Pull request by zaqqwerty**
Resolves #490 .  Currently the cirq density matrix simulator does not support this interface, see https://github.com/quantumlib/Cirq/issues/3964, so there are switches to still accept that sim.  See also https://github.com/quantumlib/Cirq/pull/3927, we can switch to the public import at some point in the future.

### CODE REVIEW

File: tensorflow_quantum/core/ops/batch_util.py [hunk 1]
```diff
@@ -451,45 +428,30 @@ def batch_calculate_expectation(circuits, param_resolvers, ops, simulator):
...
+                    sum(
+                        term._expectation_from_density_matrix_no_validation(
+                            dm, qubit_order) for term in op) for op in o
+                ])
+            else:
+                # Valid observables always have real expectation values.
+                all_exp_vals.append(
+                    simulator.simulate_expectation_values(c, o, p))
+
+    return np.real(np.asarray(all_exp_vals)).astype(np.float32)
```
Comment (MichaelBroughton) [inline tensorflow_quantum/core/ops/batch_util.py:454]: This removes the parallel component here. I'm wondering if maybe we should remove the parallel component for all of our python backends (it's a fair bit more trouble than it's worth). We aren't really using them for much these days. Thoughts ?

**Review by MichaelBroughton — COMMENTED** [review]
Looks great, just wondering what we should do about the parallel operations in `batch_util.py`

File: tensorflow_quantum/core/ops/batch_util.py (hunk 1 above)
Comment (zaqqwerty) [reply tensorflow_quantum/core/ops/batch_util.py:454]: Yeah I think the python backends should make their own choices about parallel simulation.  Using just a simple loop, we can accommodate any style of cirq object, whether it is local, remote, sim, or real device.  If we want real batching over circuits I think we should have the cirq objects support it instead.

Comment (MichaelBroughton) [reply tensorflow_quantum/core/ops/batch_util.py:454]: Good call. Would you be able to open a new PR that rips out all the parallelization after we merge this ?

Comment (zaqqwerty) [reply tensorflow_quantum/core/ops/batch_util.py:454]: Yeah I can do that, opened this issue: #517

File: tensorflow_quantum/core/ops/batch_util.py [hunk 2]
```diff
@@ -451,45 +428,30 @@ def batch_calculate_expectation(circuits, param_resolvers, ops, simulator):
...
+                qubit_order = dict(
+                    zip(sorted(c.all_qubits()),
+                        list(range(len(c.all_qubits())))))
+                sim_result = simulator.simulate(c, p)
+                dm = sim_result.final_density_matrix
+                all_exp_vals.append([
+                    sum(
+                        term._expectation_from_density_matrix_no_validation(
+                            dm, qubit_order) for term in op) for op in o
+                ])
```
Comment (MichaelBroughton) [inline tensorflow_quantum/core/ops/batch_util.py:448]: I think this is doing unnecessary work and can be simplified to just:
```python3
for j, sub_op in enumerate(o):
  all_exp_vals[i][j] = sub_op.expectation_from_density_matrix_no_validation(dm, qubit_order)
```

**Review by MichaelBroughton — COMMENTED** [review]
Looking good just a few simplifications in the DM part and we should be good to merge.

File: tensorflow_quantum/core/ops/batch_util.py [hunk 3]
```diff
@@ -451,45 +428,30 @@ def batch_calculate_expectation(circuits, param_resolvers, ops, simulator):
...
-    with ProcessPool(processes=None,
-                     initializer=_setup_dict,
-                     initargs=(shared_array, return_mem_shape, simulator,
-                               post_process)) as pool:
-
-        pool.starmap(_analytical_expectation_worker_func, input_args)
-
-    return _convert_simple_view_to_result(shared_array, np.float32,
-                                          return_mem_shape)
+    all_exp_vals = []
```
[the remark this answers is not in the document]
Comment (zaqqwerty) [reply tensorflow_quantum/core/ops/batch_util.py:431]: Done

File: tensorflow_quantum/core/ops/batch_util.py [hunk 4]
```diff
@@ -451,45 +428,30 @@ def batch_calculate_expectation(circuits, param_resolvers, ops, simulator):
...
+    for c, p, o in zip(circuits, param_resolvers, ops):
+        # Convention in TFQ is to set expectations of empty circuits to -2.
+        if len(c) == 0:
+            all_exp_vals.append(np.full(len(o), -2.0, dtype=np.float32))
```
[the remark this answers is not in the document]
Comment (zaqqwerty) [reply tensorflow_quantum/core/ops/batch_util.py:435]: Done

File: tensorflow_quantum/core/ops/batch_util.py [hunk 5]
```diff
@@ -451,45 +428,30 @@ def batch_calculate_expectation(circuits, param_resolvers, ops, simulator):
...
+        else:
+            # TODO(zaqqwerty): remove DM sim check once cirq #3964 is resolved.
+            if isinstance(simulator, cirq.DensityMatrixSimulator):
```
[the remark this answers is not in the document]
Comment (zaqqwerty) [reply tensorflow_quantum/core/ops/batch_util.py:438]: Done

File: tensorflow_quantum/core/ops/batch_util.py [hunk 6]
```diff
@@ -451,45 +428,30 @@ def batch_calculate_expectation(circuits, param_resolvers, ops, simulator):
...
+                qubit_order = dict(
+                    zip(sorted(c.all_qubits()),
+                        list(range(len(c.all_qubits())))))
```
[the remark this answers is not in the document]
Comment (zaqqwerty) [reply tensorflow_quantum/core/ops/batch_util.py:441]: Done

File: tensorflow_quantum/core/ops/batch_util.py (hunk 2 above)
Comment (zaqqwerty) [reply tensorflow_quantum/core/ops/batch_util.py:448]: Done

**Review by MichaelBroughton — APPROVED** [review]
LGTM

**Pull request merged**